### PR TITLE
Partial drawing of any Object

### PR DESCRIPTION
**Pull request by ArbitRandomUser**
## PR Checklist

If you are contributing to `Javis.jl`, please make sure you are able to check off each item on this list:

- [x] Did I update `CHANGELOG.md` with whatever changes/features I added with this PR?
will add later
- [x] Did I make sure to only change the part of the file where I introduced a new change/feature?
- [x] (Hopefully)  Did I cover all corner cases to be close to 100% test coverage (if applicable)?
- [x] Did I properly add Javis dependencies to the `Project.toml` + set an upper bound of the dependency (if applicable)?
- [x] (NA) Did I properly add test dependencies to the `test` directory (if applicable)?
- [x] (NA) Did I check relevant tutorials that may be affected by changes in this PR?
- [x] Did I clearly articulate why this PR was made the way it was and how it was made?

**Link to relevant issue(s)**
Closes #
https://github.com/JuliaAnimators/Javis.jl/issues/218

**How did you address these issues with this PR? What methods did you use?**

This PR introduces the "show_creation" function for `Action` ; 
Using Cassette we intercept  any call to strokepath , strokepreserve , fillpath and fillpreserve .
they are replaced with custom set of strokes that stroke the path/fill upto a particular fraction.

for example , this bit of code makes ...
```julia
using Javis
using Animations
using LaTeXStrings

video=Video(600,600)

function circle_line()
    sethue("red")
    circle(O,100,:stroke)
    sethue("green")
    line(O,O+100,:stroke)
    sethue("blue")
    rect(O-100,50,50,:fill)
    sethue("orange")
    rect(O+100,60,90,:path)
    #the usuall fillstroke calls fill first then stroke
    #the partial drawing happens in the order you call
    #these luxor functions. Since i want the outline
    #first then the fill...
    strokepreserve()
    fillpath()
end

nframes=200
exf=30
Background(1:nframes+exf, (args...)-> background("black"))
obj = Object(1:nframes+exf, (_,_,_)-> ( circle_line());)
act!(obj,Action(1:nframes,linear(),show_creation()))
render(video, pathname="circle_line_rect.gif")
```

will give ...
circle_line_rect

### CODE REVIEW

File: src/draw_partial.jl [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+#to stop furthur drawing from happening
+
+"""A slightly modified version of pathtopoly ,
+by default pathtopoly() will dispatch to Luxors pathtopoly,
+however passing any symbol will dispatch to the following
+function. Main difference is a new move command starts
+a new subpath, we also return if each subpath is closed or
+open in an Array{Bool},
+"""
+function Luxor.pathtopoly(co_state::Symbol)
```
Comment (Wikunia) [inline src/draw_partial.jl:18]: The function name and parameters should be the first line of each docstring. Feel free to check others to see the general style we use.

File: Project.toml [hunk 2]
```diff
@@ -6,6 +6,7 @@ version = "0.8.0"
 [deps]
 Animations = "27a7e980-b3e6-11e9-2bcd-0b925532e340"
 Cairo = "159f3aea-2a34-519c-b102-8c37f9878175"
+Cassette = "7057c7e9-c182-5462-911a-8362d720325c"
```
Comment (Wikunia) [inline Project.toml:9]: Needs also a compat entry for later ☺️

**Comment by ArbitRandomUser** [comment]
i'm thinking of writing some tests ,
1) a the animation above in the P.R .
2) scaling, translating, rotating and fading along with show_creation
3) anything else ???
@Wikunia  @gpucce  @TheCedarPrince  @Sov-trotter 
let me know your comments (if any) on other things i should test

**Comment by Sov-trotter** [comment]
This looks really solid!
I have one question, can we make it work with the `JObject` syntax?

**Comment by ArbitRandomUser** [comment]
> This looks really solid! I have one question, can we make it work with the `JObject` syntax?

if you meant something like this ...
```julia
using Javis
using Animations

video=Video(600,600)
nframes=100
Background(1:nframes, (args...)-> background("black"))

obj = Object(JCircle(O,100,color="white"))
obj2= Object(JRect(O+100,60.0,90.0,color="orange",action=:fill))

act!(obj,Action(1:nframes,linear(),show_creation()))
act!(obj2,Action(1:nframes,linear(),show_creation()))
render(video, pathname="circle_line_rect.gif")
``` 
yup it works outta the box ! 
circle_line_rect

File: Project.toml (hunk 2 above)
Comment (ArbitRandomUser) [reply Project.toml:9]: is this manually added ? or is the something i have to do from Pkg?

**Comment by ArbitRandomUser** [comment]
i've commited the intermediate refs images , should i commit the gifs that are generated also ? 

circle_move_creation
circle_scale_creation
rect_rotate_creation
show_creation

**Comment by Wikunia** [comment]
Currently we only test reference images and I think that works quite well. If there is anything you can test without images it's always appreciate but often it can't be done.

File: src/draw_partial.jl [hunk 3]
```diff
@@ -0,0 +1,474 @@
...
+            if dp_state.cur_len_partial + nextpolylength < dp_state.target_len_partial
+                poly(poly_i, close = co)
+                dp_state.cur_len_partial += nextpolylength
+            else
+                if length(poly_i) >= 2
+                    frac =
+                        (dp_state.target_len_partial - dp_state.cur_len_partial) /
+                        nextpolylength
+                    poly(polyportion(poly_i, frac, closed = co, pdist = pdist_i))
+                    #poly(polyportion(poly_i, frac, closed = co))
```
Comment (ArbitRandomUser) [inline src/draw_partial.jl:273]: there seems to be no difference in performance if i switch this line for the line above (with `pdist = pdist_i`) ;
atleast for the example i tried (hilbert curve)
we would expect passing the polydistances to polyportion to make it faster .

File: src/draw_partial.jl [hunk 4]
```diff
@@ -0,0 +1,474 @@
...
+        push!(dp_state.cur_len_perim, distance(bbox[1], bbox[2]))
+    else
+        push!(dp_state.cur_len_perim, 0)
+    end
+end
+
+#hue
+#="""for some reason parsing colors fails to furthur overdub , so we just
+return as is"""=#
+function overdub(c::ctx_strokelength, ::typeof(Colors.parse), args...)
```
Comment (ArbitRandomUser) [inline src/draw_partial.jl:151]: Colors.parse simply doesn't play well with Cassette , might be something related to this https://github.com/JuliaLabs/Cassette.jl/issues/130

**Comment by Ved-Mahajan** [comment]
WOW, THIS IS ABSOLUTELY BEAUTIFUL!!!

**Comment by ArbitRandomUser** [comment]
merged in #482
